src/uct/tcp/tcp_ep.c

```diff
@@ -499,28 +499,27 @@ static ucs_status_t uct_tcp_ep_connect(uct_tcp_ep_t *ep)
     uct_tcp_ep_t *peer_ep  = NULL;
     ucs_status_t status;
 
-    /* It checks that the connect was issued:
-     * - from uct_ep_connect_to_ep(), if the EP created using CONNECT_TO_EP
-     *   method and a local side must connect to a peer
-     * - from uct_ep_create(), if the EP created using CONNECT_TO_IFACE method
+    /* Check that the connection was not issued yet. New connection has to be
+     * issue form:
+     * - uct_ep_connect_to_ep(), if the EP created using CONNECT_TO_EP method
+     *   and a local side must connect to a peer (due to address resolution
+     *   logic)
+     * - uct_ep_create(), if the EP created using CONNECT_TO_IFACE method
      *   and no an internal EP was created due to connection from a peer */
     ucs_assert((ep->conn_state == UCT_TCP_EP_CONN_STATE_CLOSED) &&
                (ep->conn_retries == 0));
 
     uct_tcp_cm_change_conn_state(ep, UCT_TCP_EP_CONN_STATE_CONNECTING);
 
-    if (!uct_tcp_ep_is_self(ep) &&
-        !(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP)) {
-        peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->cm_id.conn_sn,
-                                    UCT_TCP_EP_FLAG_CTX_TYPE_RX);
+    if (uct_tcp_ep_is_self(ep) ||
+        (ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP)) {
+        goto create_and_connect_out;
     }
 
+    peer_ep = uct_tcp_cm_get_ep(iface, &ep->peer_addr, ep->cm_id.conn_sn,
+                                UCT_TCP_EP_FLAG_CTX_TYPE_RX);
     if (peer_ep == NULL) {
-        status = uct_tcp_ep_create_socket_and_connect(ep);
-        if (status != UCS_OK) {
-            return status;
-        }
-        goto out;
+        goto create_and_connect_out;
```

Comment:
maybe like this:
if (peer_ep == NULL) {
  status = uct_tcp_ep_create_socket_and_connect(ep);
} else {
 /* all the code in lines 525-543 */
}

Reply:
fixed